Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp

```diff
@@ -26,59 +26,137 @@ using namespace AZ::Internal;
+    uint64_t pack() const
     {
-        result = m_queue.front();
-        m_queue.pop_front();
+        uint64_t out;
+        memcpy(&out, this, sizeof(out));
+        return out;
     }
 
-    return result;
-}
+    [[nodiscard]] bool empty() const
+    {
+        return head == tail;
+    }
+
+    // Typical circular buffer pattern where the head chases the tail. For two phase commits, tailReserve can run ahead of tail
+    // but can't bump into head. Note that for all operations here, we rely on well-defined wrapping behavior of unsigned ints.
+    [[nodiscard]] bool canEnqueue() const
+    {
+        return tailReserve != head - 1;
+    }
+
+    void reserve()
+    {
+        ++tailReserve;
+    }
 
-Job* WorkQueue::TryStealFront()
+    void commit()
+    {
+        ++tail;
+    }
+
+    void acquire()
+    {
+        ++head;
+    }
+
+    uint16_t head;
+    uint16_t tail;
+    uint16_t tailReserve;
+    uint16_t unused;
+};
+
+void JobQueue::Enqueue(Job* job)
 {
-    AZStd::exponential_backoff backoff;
-    for (unsigned attempCount = 0; attempCount < TryStealSpinAttemps; ++attempCount)
+    uint8_t priority = job->GetPriority();
+    uint64_t previousStatus = m_status[priority].load(AZStd::memory_order_acquire);
+
+    while (true)
     {
-        // Do a bounded spin with backoff to acquire the lock
-        if (m_lock.try_lock())
+        StatusPayload status{ previousStatus };
+
+        // Enqueuing is done in two phases because we cannot atomically write the job to the slot we reserve
+        // and simulataneously publish the fact that the slot is now available.
+        if (status.canEnqueue())
         {
-            Job* result = nullptr;
-            if (!m_queue.empty())
+            // Try to reserve a slot
+            status.reserve();
+            uint64_t reserveStatus = status.pack();
+            if (m_status[priority].compare_exchange_weak(previousStatus, reserveStatus))
             {
-                result = m_queue.front();
-                m_queue.pop_front();
+                m_queues[priority][status.tailReserve - 1] = job;
+
+                status.commit();
+
+                while (true)
+                {
+                    // Increment the tail to advertise the new job
+                    if (m_status[priority].compare_exchange_weak(reserveStatus, status.pack()))
+                    {
+                        return;
+                    }
+
+                    status = reserveStatus;
+                    status.commit();
+                }
             }
 
-            m_lock.unlock();
-            return result;
+            // We failed to reserve a slot, try again
+        }
+        else
+        {
+            // All queues at or below the requested priority are full. Queue in the global unbounded queue instead.
+            job->GetContext()->GetJobManager().QueueUnbounded(job);
+            job->GetContext()->GetJobManager().ActivateWorker();
+            return;
         }
+    }
+}
```

Comment:
Very rarely will this else block gets hit. There would have to be 65534 jobs queued in order for it get hit.
We should add a test to make sure this control path actually works when he queue is full.

Reply:
Do the benchmark tests not get run as part of Jenkins?

Reply:
Yes, Benchmark test sare run though the Benchmarks jobs for the Development branch Nightly Job on Jenkins(they don't run as part of AR).

I just want to make sure that we have at least tested this code path in case there are potential issues with it.

Reply:
Even for the Benchmark Jobs, the maximum number of jobs are set at [16384](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/Tests/Jobs.cpp#L1705).

Reply:
Not ignoring this, just thinking of the best way to test it

Reply:
FYI I have a test but it's... flaky and unfortunately appears to be an existing issue. At high volume I'm seeing missed notifications/wakeups on the global queue but this seems to be true regardless of where the job is queued from (this code or the previous code). My recommendation is for us to proceed here. I'm going to add a test for this in this PR but leave it disabled and create an issue for it later.

Reply:
That is good with me.